sender.go

```diff
@@ -69,32 +74,58 @@ func (h *Hub) newSender(ctx context.Context) (*sender, error) {
 			Max:    4 * time.Second,
 			Jitter: true,
 		},
+		cond: sync.NewCond(&sync.Mutex{}),
 	}
 	tab.For(ctx).Debug(fmt.Sprintf("creating a new sender for entity path %s", s.getAddress()))
 	err := s.newSessionAndLink(ctx)
 	return s, err
 }
 
+func (s *sender) amqpSender() *amqp.Sender {
+	return s.sender.Load().(*amqp.Sender)
+}
+
 // Recover will attempt to close the current session and link, then rebuild them
 func (s *sender) Recover(ctx context.Context) error {
 	span, ctx := s.startProducerSpanFromContext(ctx, "eh.sender.Recover")
 	defer span.End()
-
-	// we expect the sender, session or client is in an error state, ignore errors
-	closeCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
-	defer cancel()
-	_ = s.sender.Close(closeCtx)
-	_ = s.session.Close(closeCtx)
-	_ = s.connection.Close()
-	return s.newSessionAndLink(ctx)
+	recover := false
+	// acquire exclusive lock to see if this goroutine should recover
+	s.cond.L.Lock()
+	if !s.recovering {
+		// another goroutine isn't recovering, so this one will
+		s.recovering = true
+		recover = true
+	} else {
+		// wait for the recovery to finish
+		s.cond.Wait()
+	}
+	s.cond.L.Unlock()
+	var err error
+	if recover {
+		// we expect the sender, session or client is in an error state, ignore errors
+		closeCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+		// update shared state
+		s.cond.L.Lock()
```

Comment:
Looking at this I see a potential scenario but I'm not entirely sure about the surrounding circumstances so it may be impossible. 

For the purposes of this I'll call the first block of code protected by `s.cond.L.Lock()` calls 'block 1' and the second `s.cond.L.Lock()` call that occurs inside of the `if (recover)` statement `block 2`:

1. Several goroutines are attempting to recover
2. goroutine A starts and is the first, so it locks, sets `s.recovering = true, recover = true`, enabling the behavior in block 2 for itself.
3. other goroutines have not yet entered into `Recover`
4. goroutine A now starts into block2, again, locking `s.cond.L.Lock()`
5. at this point other goroutines now attempt to recover. They are paused on Block 1.
6. 'A' now completes recovery, s.recovering to false, unlocks and broadcasts
7. All the other goroutines waiting on block 1 come in, and now also attempt to recover 
8. At least one of the additional goroutines will try to recover again since s.recovering was set to false (in step 6, by goroutine A). 

Is this a possibility?

Reply:
Yes I believe so.  It won't be susceptible to a data-race, but it does mean that we needlessly recover the connection.

I don't see a way to fix this without using a `RWLock` to block sending messages while a recover is in progress, however I worry that it might hurt scaleability.  I think we'd need to measure the impact such a change would have before deploying it, and I don't think we have a good way to do that at present.

Reply:
I think it's an acceptable compromise, TBH (at the very least, it solves the problem in a way that's non-damaging). Can we add in a comment that this is a known scenario?

Reply:
Added comment explaining the current limitation.